src/lib/sql/sql_pipeline_statement.hpp

```diff
@@ -103,6 +105,11 @@ class SQLPipelineStatement : public Noncopyable {
   const std::shared_ptr<SQLPhysicalPlanCache> pqp_cache;
   const std::shared_ptr<SQLLogicalPlanCache> lqp_cache;
 
+  static void expression_parameter_extraction(std::shared_ptr<AbstractExpression>& expression,
+                                              std::vector<std::shared_ptr<AbstractExpression>>& values);
+  const std::shared_ptr<AbstractLQPNode> split_logical_plan(const std::shared_ptr<const AbstractLQPNode>& logical_plan,
+                                                            std::vector<std::shared_ptr<AbstractExpression>>& values);
+
```

Comment:
I wonder if the SQLPipelineStatement is really the best place for this. It appears to me as if we are moving LQP modification logic into a class that was previously only concerned with executing a single statement of an SQL query.

Maybe (!) it makes sense to have a component that deals with the "generalization" of queries (i.e., extracting and replacing the value expressions) as well as with applying extracted value expressions to a previously generalized LQP.

Haven't looked at the code in the CPP too much yet. We probably should discuss the architecture before we dive into details there.

Reply:
It has been some time since this came up but eventually @Bensk1 and I discussed the possible implementations of this.
The idea is now that we implement an facade/adaptor like class for the cache that would basically implement try_get and set methods and handle all the extracting and replacement as you mentioned.
However, this requires that this class holds state between try_get and set since the extracted values and the unoptimized lqp without values need to be kept for later instantiation of the optimized plan and as cache key. With multithreading in mind, this requires that we set up a new adaptor object for each new pair of try_get and set operations. Even though this might create a little overhead in terms of object creation, we think that this is the best version. The other options would be keeping the state in the sql_pipeline_statement but moving the logic/functionality to utility functions or moving the state into the cache but then we have to implement some kind of transaction context to identify related get and set calls. any objections @mrks?

Reply:
I will have to dive deeper into things before I can comment on this. It's on my list and will hopefully get done this weekend.

Reply:
Out of interest, I visualized the signal-to-noise ratio here. Green is the actual work on optimizing the plan, red is everything that has to do with caching, and blue is timing code.

![Untitled](https://user-images.githubusercontent.com/575106/104811256-9a791880-57fa-11eb-9298-16753a5fef5e.png)

I don't have a good idea on what to do with the timing, either, but it bugs me.

As for your suggestion. Do I understand you correctly in that it would basically look something like this?

```cpp
class DeparameterizedLQPCacheHandler {
  DeparameterizedLQPCacheHandler(const std::shared_ptr<SQLLogicalPlanCache>& cache) : cache(cache) {}

  std::optional<std::shared_ptr<AbstractLQPNode>> try_get(const std::shared_ptr<AbstractLQPNode>& unoptimized_lqp) {
    Assert(!deparameterized_unoptimized_plan, "can only be used once");

    auto [tmp, extracted_values] = split_logical_plan(unoptimized_lqp);
    deparameterized_unoptimized_plan = tmp;

    auto deparameterized_optimized_plan = cache->try_get(deparameterized_unoptimized_plan);
    if (!deparameterized_optimized_plan) return std::nullopt;
    return instantiate(deparameterized_optimized_plan, extracted_values);
  }

  void set(const std::shared_ptr<AbstractLQPNode>& optimized_lqp) {
    Assert(deparameterized_unoptimized_plan, "must use try_get first");
    auto [deparameterized_optimized_plan, extracted_values] = split_logical_plan(optimized_lqp);
    cache->set(deparameterized_unoptimized_plan, deparameterized_optimized_plan);
  }
    
  std::shared_ptr<SQLLogicalPlanCache> cache;
  std::shared_ptr<AbstractLQPNode> deparameterized_unoptimized_plan;
};
```

(Names tbd. of course. Don't use DeparameterizedLQPCacheHandler, please.)

If this comes close to what you were proposing, this sounds good. I don't see how we create any object that we would not have created otherwise, given that the `DeparameterizedLQPCacheHandler` instance is just a stack variable.

This should make it much easier to test, too.

If I did not understood your proposal, could you please correct my pseudo code?

Reply:
Your draft is pretty much what i had in mind. i would store the extracted values from the try get to get them in the right order in set rather than the deparameterized_unoptimized_plan.

how did you highlighted the different code sections? by hand or is there a tool for this?

Concerning the naming: you already mentioned that the difference between PreparedPlan and deparameterized/placeholder whatever LQP is not very clear. In the end it is nothing but a prepared plan that is cached and our "Handler" (if we want to call it so) would basically convert LQPs into PreparedPlans and back. So what about dropping the term of "deparameterized" and just saying that we are dealing with PreparedPlans on the cache side and LQPs on the sql pipeline statement side and the new class is the ```PreparedPlanCacheHandler```?

Reply:
> how did you highlighted the different code sections? by hand or is there a tool for this?

By hand.

> Concerning the naming: [...]

To me, "prepared plan" sounds quite close to SQL's `PREPARE`. Have you checked the papers that discuss \emph{safe plan reuse} if they use a common terminology?

Reply:
Did I just type `\emph` in Github? It's really time to finish this thesis...

Reply:
> It's really time to finish this thesis...

Bitte zügeln Sie Ihre Zunge!

Reply:
most of the caching logic and cache access timing is now in ParameterizedPlanCacheHandler (parameterized plan seems to be used in the literature). however, we still have the basic cache access and optimization timing in this function.